src/libraries/System.Security.Cryptography.Cose/tests/CoseHeaderMapTests.cs

```diff
@@ -0,0 +1,248 @@
+
+            static ReadOnlyMemory<byte> EncodeInt32(int value, CborWriter writer)
+            {
+                writer.WriteInt32(value);
+                return EncodeAndReset(writer);
+            }
+
+            static ReadOnlyMemory<byte> EncodeString(string value, CborWriter writer)
+            {
+                writer.WriteTextString(value);
+                return EncodeAndReset(writer);
+            }
+
+            static ReadOnlyMemory<byte> EncodeBytes(ReadOnlySpan<byte> value, CborWriter writer)
+            {
+                writer.WriteByteString(value);
+                return EncodeAndReset(writer);
+            }
+
+            static ReadOnlyMemory<byte> EncodeAndReset(CborWriter writer)
+            {
+                ReadOnlyMemory<byte> encodedValue = writer.Encode();
+                writer.Reset();
+                return encodedValue;
+            }
+        }
+
+        [Fact]
+        public void DecodedProtectedMapShouldBeReadOnly()
+        {
+            byte[] encodedMessage = CoseSign1Message.Sign(s_SampleContent, ECDsaKeys[ECDsaAlgorithm.ES256], HashAlgorithmName.SHA256);
+            CoseSign1Message message = CoseMessage.DecodeSign1(encodedMessage);
+            Assert.True(message.ProtectedHeader.IsReadOnly);
+        }
+
+        [Fact]
+        public void GetValueFromReadOnlyProtectedMap()
+        {
+            byte[] encodedMessage = CoseSign1Message.Sign(s_SampleContent, ECDsaKeys[ECDsaAlgorithm.ES256], HashAlgorithmName.SHA256);
+            CoseSign1Message message = CoseMessage.DecodeSign1(encodedMessage);
+            Assert.True(message.ProtectedHeader.IsReadOnly);
+
+            int expectedAlgorithm = (int)ECDsaAlgorithm.ES256;
+
+            int algorithm = message.ProtectedHeader.GetValueAsInt32(CoseHeaderLabel.Algorithm);
+            Assert.Equal(expectedAlgorithm, algorithm);
+
+            ReadOnlyMemory<byte> encodedAlgorithm = message.ProtectedHeader.GetEncodedValue(CoseHeaderLabel.Algorithm);
+            Assert.Equal(expectedAlgorithm, new CborReader(encodedAlgorithm).ReadInt32());
+
+            message.ProtectedHeader.TryGetEncodedValue(CoseHeaderLabel.Algorithm, out encodedAlgorithm);
+            Assert.Equal(expectedAlgorithm, new CborReader(encodedAlgorithm).ReadInt32());
+        }
+
+        [Fact]
+        public void SetValueAndRemoveThrowIfProtectedMapIsReadOnly()
+        {
+            byte[] encodedMessage = CoseSign1Message.Sign(s_SampleContent, ECDsaKeys[ECDsaAlgorithm.ES256], HashAlgorithmName.SHA256);
+            CoseSign1Message message = CoseMessage.DecodeSign1(encodedMessage);
+            Assert.True(message.ProtectedHeader.IsReadOnly);
+
+            var barLabel = new CoseHeaderLabel("bar");
+            Assert.Throws<InvalidOperationException>(() => message.ProtectedHeader.SetValue(barLabel, 42));
+            Assert.Throws<InvalidOperationException>(() => message.ProtectedHeader.Remove(barLabel));
```

Comment:
Should also include a value that was present.  e.g. try to overwrite the AlgorithmId and then remove it.

Yeah, from reading the code I know that doesn't matter, but from a black-box test perspective they're two different things to check.